prevent multiple orchestrations at the same time

shaas opened the pull request:
Fixes #2858 
Signed-off-by: Stefan Haas <shaas@suse.com>

<!-- Please take a look at our [Contributing](https://rook.io/docs/rook/master/development-flow.html)
documentation before submitting a Pull Request!
Thank you for contributing to Rook! -->

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #2858 

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

// for known CI issue
[skip ci]

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -291,47 +282,22 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
-	defer c.orchRunMux.Unlock()
-	c.orchRunMux.Lock()
+func (c *cluster) checkSetOrchestrationStatus() bool {
+	defer c.orchMux.Unlock()
```
This `defer` must be placed after the lock. If something were to go wrong and the function tried to unlock an already unlocked mutex, there is a panic: `fatal error: sync: unlock of unlocked mutex`

shaas replied:
Done

shaas commented:
Please note that I did additional changes to the code as the previous commit was not correct.

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -290,47 +291,37 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
+func (c *cluster) unsetOrchestrationStatus() {
+	c.orchMux.Lock()
+	c.orchestrationRunning = false
+	c.orchestrationPending = true
```
Why does completion of the orchestration running imply orchestration is pending?

shaas replied:
With `orchestrationPending` the `for-loop` in `createInstance` gets controlled. It is responsible to 
1) get into the loop (that is the reason why the variable needs to get initialized and reset to `true`) and
2) get the loop running as long as there is a new request pending.

Well, the name of this variable is somehow misleading. Any suggestions for a better name?

BlaineEXE replied:
So is the role more something like `readyForOrchestrationRun`?

shaas replied:
Well, not really. Because `orchestrationPending` set does not mean it is ready for orchestration (e.g. there is already an orchestration running).

BlaineEXE replied:
I think maybe my initial question didn't make sense. Why does the `unsetOrchestrationStatus` method, which I presume is called when an orchestration is finished running, set `orchestrationPending`? This seems like it will create a situation where orchestrations are infinitely pending, even when there are no updates. I would think that the only thing that should set `orchestrationPending = true` would be an event that triggers an orchestration run.

shaas replied:
`unsetOrchestrationStatus` gets called when `createInstance` is done -> the for-loop which does the actual orchestration does not get executed anymore. The next orchestration run will start as soon as  `createInstance` gets called e.g. if a node got added or the CRD got altered. Resetting ist needed `orchestrationPending` as this is one of the two criterias which allows to enter the for-loop (orchestration).
1) There is no other orchestration run currently (gets checked at the beginning of 'createInstance`)
2) `orchestrationPending` is set to true. This means that it has to be initialized as `true` and it needs to be reset to `true` when _all_ orchestration-runs are done otherwise the `for`-loop is not "enterable". Again, the name is misleading. But I have not better idea ;)

BlaineEXE commented:
Please also update the commit header to begin with `ceph:`

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -138,16 +138,17 @@ func (c *cluster) detectCephVersion(image string, timeout time.Duration) (*cephv
-	initRun := true
-	for c.checkUnsetOrchestrationPending() == true || initRun == true {
-		initRun = false
+	defer c.unsetOrchestrationStatus()
```
I don't think these actions are atomic enough. There is a race condition possible still (though very unlikely) where the for loop checks `c.checkSetOrchestrationStatus()`, which returns false, at which point the scheduler pauses this routine. About this time, another call to `createInstance` is made, and `c.getOrchestrationRunning` returns true, causing `c.setOrchestrationPending()` to be called. The scheduler continues with the paused routine, which returns nil and executes the `defer`, setting `orchestrationRunning = false` and `orchestrationPending = true`. But now the operator *should* be updating the cluster, but it won't until another update is made (or the operator is restarted), resulting in a new call to `createInstance`.

I think the for loop logic needs to be totally separated from the `createInstance` method. What about this flow ... ?
1. Start a go routine that's purpose is to check if `orchestrationPending == true`, and if so, set `orchestrationRunning = true`, and then call the original `createInstance`. Once `createInstance` returns, `orchestrationRunning` should be set back to `false`. If another orchestration has been made `pending` during the orchestration run, the loop will see that on the next iteration.
2. Instead of calling `createInstance` from the detection of a cluster change event, simply call a method to set `orchestrationPending = true` and return. This may have to update a stored version of the latest pending `rookImage` and `cephVersion` for item 1 above.
3. On operator startup, `orchestrationPending` should still be set to `true`.

I think with this flow, there will be no misleading variable names, and the possible race condition will be eliminated.

shaas replied:
I had also the idea with a separate routine for `createInstance`. But while checking the code I had to notice that there are subsequent calls which rely on the results of `createInstance`.
This means that this solution is not easily possible.
Therefore I moved all get and set calls of the orchestration variables to `c.checkSetOrchestrationStatus()`. No race conditions should be possible anymore.

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -290,48 +289,33 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
-	if c.orchestrationPending == true {
+		// allow to run an orchestration
+		return true
+	} else if c.orchestrationPending == true && c.orchestrationRunning == true && init == true {
```
This condition seems dangerous (and unused?). If an orchestration is pending, and another thread is running orchestration, and this is not the first loop, go ahead and run an orchestration anyway?

shaas replied:
This condition is used by the running orchestration thread (-> `init` is set) to indicate that this is the initial `createInstance` call.



| Thread A | Thread B | env | comment |
| ------------ |------------ | -------------- | ------------ |
| createInstance| | pend = T, run = F | |
| for true (pend = T & run = F, init = false) | | | |
| set init = true | | pend = F, run = T | init is thread-local|
| _work_ | createInstance | pend = F, run = T | |
| _work_ | for false (pend = F & run = T, init = false) | pend = T, run = T | Thread B set pend|
| for true (pend = T & run = T, init = true) | | pend = F, run = T | Thread A starts **another** loop | as it is the initial `createInstance`-Instance and as there is a pending orchestration |

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -66,13 +65,14 @@ func newCluster(c *cephv1.CephCluster, context *clusterd.Context) *cluster {
+		context:              context,
+		stopCh:               make(chan struct{}),
+		ownerRef:             ownerRef,
+		orchestrationPending: true,
```
Could we get a comment here to explain that this is initialized to true so that the first call to orchestrate the cluster will go through?

BlaineEXE replied:
I think it's unnecessary to initialize this to true anyway. Any call to `createInstance` (or `handleUpdate` automatically implies that at least one update is pending, the one that spawned the call to the method. Upon entry to the method, I would set (atomically) orchestration pending. (Atomically) If there is already an orchestration running, the function can return without error. (Atomically) If there is not an orchestration running and there is an orchestration pending, then set running true and pending false. Deep copy the spec. Once the cluster is updated, (atomically) set orchestration running false. The loop will check again (atomically) if there is an orchestration pending and if there is not an orchestration running (another thread may have taken over the pending update).

BlaineEXE replied:
```go

func updateWork() retvals {
    // work
}

func updateStuff(...) {
    pending = true
    for atomic(pending==true && running==false, running=true, pending=false) {
        // deep copy spec
        retvals = func() {
            defer atomic(running=false)
            return updateWork()
        }
    }
}
```

shaas replied:
Setting `c.orchestrationPending` here make sense. I will remove the initialization.

shaas commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -138,16 +138,15 @@ func (c *cluster) detectCephVersion(image string, timeout time.Duration) (*cephv
-		c.setOrchestrationPending()
-		return nil
-	}
+	init := false
```
This variable init is described here: pkg/operator/ceph/cluster/cluster.go:302
"// the thread local variable init indicates if the calling function is the initial one which started the
// orchestration."

shaas commented:
> The short here is that I'm still mostly confused about what is going on.
> 
> I am still having trouble understanding what is the intention with the `orchPending` variable. The name is misleading, and I think that we should make it a point not to mislead future devs. This might mean revisiting the problem from the beginning. It seems to me that `orchPending` may represent the logical or-ing of 2 (or more?) separate conditions. I think it might being taken to mean `has an orchestration been requested?` and `are all other orchestrations complete?`.
> 
> I see in the code that the result of the call to `createInstance` is used. One is in `onK8sNodeAdd`, which you added recently. Another is in `onK8sNodeUpdate`, which you added at the same time. The other usages are `onAdd` and in the method `handleUpdate`. `handleUpdate` is called in `onUpdate`, and that is it. I think maybe the k8s node changes should call `handleUpdate` instead of `createInstance` so that the cluster status also gets updated with the cluster.
> 
> I went through the code here and laid out what happens when 2 threads call `createInstance`.
> 
> thread A	thread B	Env
> pend = T, run = F
> createInstance with Spec1		
> init = 0		
> for true (b/c pend = T & run = F)		pend = F, run = T
> init = 1		
> deep copy Spec1		
> _work_		
> _work_		
> _work_	createInstance with Spec2	
> "	init = 0	
> "	for false (b/c pend=F & run=T)	
> "	return nil	
> "		
> for false (b/c pend=F & run=T)		
> return nil		pend=T, run=F
> Unless I missed something, this means that the second call to `createInstance` would immediately report success despite the fact that spec2 is not going to be set. Also, in the current implementation, it doesn't seem possible for `pendinig` and `running` to be the same value; one is true means the other is false, so one of these is redundant here.
> 
> It's subtle, but I think we have to assume that once an orchestration is already running, that orchestration won't have detected the latest changes to the spec.

The second run of `createInstance` sets `c.orchestrationPending` which means the intial `createInstance` will do at least one additional loop iteration. Each loop iteration will do a new deep-copy of the spec. This means it loops until there is no newer version of spec available.

shaas commented:
I tried to simplify the code by renaming a couple of variables and adding a bunch of comments in the code.

> There is still a race condition possible where the for loop reaches its exit condition and then, before the thread can return from the function, another thread sees that there is an orchestration running and thus sets pending=true but does not do the orchestration itself. In this case, the first thread will exit, and neither thread will have done the update needed. To fix this problem, the orchRunning status has to exist only inside of the for loop controlling the actual orchestration work.

All the checking and setting is now done in the loop and especially the `checkSetOrchestrationStatus` call.

> pending should have a single defined meaning: "should Rook update the cluster?" If pending controls entry into code paths where entry can't be explained with this defined meaning, there should be another variable. I don't believe that's necessary.

Renamed the variable to `orchestrationNeeded`. Hopefully, now it is clear what the variable stands for. This one will be set as soon as a thread calls `createInstance`

> init is still holding the opposite meaning (false means is initial, true means is not initial) that I think most people would assume from a variable named init. It is also ambiguous what "initial" as a word means in this case. I interpreted this as the initial time through the for loop, but it seems that this may instead be "the first thread that started handling the cluster update."

Renamed the variable to `isRunningThread`. It indicates if this thread is the one which runs and iterates the `for`-loop.

BlaineEXE commented:
@shaas I'm seeing this panic message at the start of the upgraded operator in my integration test logs, and it seems like it may be related to some of the node addition code. It doesn't cause the operator to crash as far as I can tell, but I'm not sure if it has some unexpected side-effects.

```
E0403 00:08:38.189642       6 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/home/blaine/development/rasputin/.tools/go-1.11/src/runtime/asm_amd64.s:522
/home/blaine/development/rasputin/.tools/go-1.11/src/runtime/panic.go:513
/home/blaine/development/rasputin/.tools/go-1.11/src/runtime/panic.go:82
/home/blaine/development/rasputin/.tools/go-1.11/src/runtime/signal_unix.go:390
/home/blaine/go/src/github.com/rook/rook/pkg/operator/ceph/cluster/controller.go:180
/home/blaine/go/src/github.com/rook/rook/pkg/operator/ceph/cluster/controller.go:138
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:195
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:314
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:444
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:444
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:150
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:124
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:124
```

shaas commented:
> @shaas I'm seeing this panic message at the start of the upgraded operator in my integration test logs, and it seems like it may be related to some of the node addition code. It doesn't cause the operator to crash as far as I can tell, but I'm not sure if it has some unexpected side-effects.
> 
> Clarification: this is in PR #2910, not in this branch.
> 
> ```
> E0403 00:08:38.189642       6 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
> /home/blaine/development/rasputin/.tools/go-1.11/src/runtime/asm_amd64.s:522
> /home/blaine/development/rasputin/.tools/go-1.11/src/runtime/panic.go:513
> /home/blaine/development/rasputin/.tools/go-1.11/src/runtime/panic.go:82
> /home/blaine/development/rasputin/.tools/go-1.11/src/runtime/signal_unix.go:390
> /home/blaine/go/src/github.com/rook/rook/pkg/operator/ceph/cluster/controller.go:180
> /home/blaine/go/src/github.com/rook/rook/pkg/operator/ceph/cluster/controller.go:138
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:195
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:314
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:444
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:444
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:150
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:124
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
> /home/blaine/go/src/github.com/rook/rook/vendor/k8s.io/client-go/tools/cache/controller.go:124
> ```

I cannot reproduce this issue with the latest code-base.

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -138,16 +137,14 @@ func (c *cluster) detectCephVersion(image string, timeout time.Duration) (*cephv
+	// isRunningThread indicates if the calling thread is the one which executes the for-loop
+	isRunningThread := false
+	c.setOrchestrationNeeded()
+
```
This would re-introduce the race condition I pointed out in the previous iteration.

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -165,23 +162,27 @@ func (c *cluster) createInstance(rookImage string, cephVersion cephver.CephVersi
 		k8sutil.SetOwnerRef(c.context.Clientset, c.Namespace, &cm.ObjectMeta, &c.ownerRef)
 		_, err := c.context.Clientset.CoreV1().ConfigMaps(c.Namespace).Create(cm)
 		if err != nil && !errors.IsAlreadyExists(err) {
+			c.unsetOrchestrationStatus()
```
`unsetOrchestrationStatus` has to be called within the context of the `for` loop to prevent a race condition where an update can be missed.

BlaineEXE replied:
Hmm ... I think you're right. This function shouldn't return until no more orchestrations are pending. I think there has to be another atomic check here. Where currently there is `c.unsetOrchestrationStatus()`, there should be something like:
```go
...
if canReturn(...) {
    return fmt.Errorf(...)
}
    logger.Errorf(... + ". there is another orchestration pending. proceeding with next orchestration run")
    continue
...
```

and where `canReturn` looks something like this
```go
func canReturn(...) bool {
    m.Lock()
    defer m.Unlock()
    if orchestrationPending {
        return false
    }
    orchestrationRunning = false
    return true
}
```

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -290,48 +296,52 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
+		// allow to run an additional iteration of the orchestration-loop
+		c.orchestrationNeeded = false
+		return true
+	} else if c.orchestrationNeeded == false && c.orchestrationRunning == true && isRunningThread == true {
```
This condition allows a second thread to have taken over orchestration work. If we are allowing another thread to take over, I think the logic in this function and in the `unset...` function are unnecessarily complicated. This can be alleviated by returning from inside the `for` loop instead of handling additional logic for when we are allowed to return after the for loop.

The below should allow another thread to take over, still reports errors to the log, and eliminates the need to follow whether this is the running thread.
```go
func orchestrate() {
    var err error
    for checkSet() {
        if err != nil {
            // there was an error encountered last loop, report it so it's not hidden
            logger.Errorf(...)
        }

        err = doStuff()
        unset()
    }

    return err
}

func checkSet() {
    m.Lock()
    defer m.Unlock()
    if needed == true && running == false {
        return true
    }
    return false
}

func unset() {
    m.Lock()
    defer m.Unlock()
    running = false
}
```

If it is not desirable for another thread to be able to take over, then the `canReturn()` method I mentioned last round would solve the issue I think.

shaas replied:
You are correct. This is with my last changes not necessary anymore.

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -290,48 +295,31 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
-	defer c.orchPenMux.Unlock()
-	return c.orchestrationPending
+// unsetOrchestrationStatus resets the orchestrationRunning-flag
+func (c *cluster) unsetOrchestrationStatus(err error) {
```
`err` is an unused parameter here.

BlaineEXE commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -138,80 +137,86 @@ func (c *cluster) detectCephVersion(image string, timeout time.Duration) (*cephv
-	initRun := true
-	for c.checkUnsetOrchestrationPending() == true || initRun == true {
-		initRun = false
+	// execute orchestration as long as there is no new orchestration request anymore
```
I am interpreting this comment in the opposite sense that it is intended. I think maybe this would be better as "execute orchestration [until] there is no new orchestration request anymore."

This seems like a good place to give an overview of the high-level intention here as well. To my summarization: "execute an orchestration until there are no more unapplied changes to the cluster definition and while no other goroutine is already running a cluster update."
I think this makes it clear in terms of the user's actions what it means when an orchestration is needed and why we need to keep track of whether an orchestration is running.
